File: 033 OOPS in C++/006 Inheritance 2.cpp
#include <bits/stdc++.h>

class Car {
    
    public:
    
    string name;
    string color;
    int engineHP; //engine horse power
    
    Car() {
        
    }
    
    Car(string name, string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP; 
    }
    
    void drive() {
        cout<<name<<" Car is driving\n";
    }
    
};

class FlyingCar : public Car {
    
    public:
    FlyingCar() {
        
    }
    
    FlyingCar(string name,string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP;
    }
    
    void fly() {
        cout<<name<<" Car is flying\n";
    }
};

class HydroFlyingCar : public FlyingCar {
    public:
    HydroFlyingCar() {
        
    }
    
    HydroFlyingCar(string name,string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP;
    }
    
    void swim() {
        cout<<name<<" Car is travelling on the surface of the water\n";
    }  
};

int main() {
    HydroFlyingCar hfc("SuperHydroFly","Black",5000);
    hfc.drive();
    hfc.fly();
    hfc.swim();
}

File: 032 OOPS in Java/006 Inheritance 2.java
import java.util.*;

class Car {
    String name;
    String color;
    int engineHP; //engine Horse Power
    
    Car(String name, String color,int engineHP) {
        this.name = name;
        this.color = color;
        this.engineHP = engineHP;
    }
    
    public void drive() {
       System.out.println("Car is driving");
    }
    
    public void reverse() {
        System.out.println("Car is in reverse mode");
    }
}

class FlyingCar extends Car {
    
    FlyingCar(String name, String color, int engineHP) {
        super(name,color,engineHP);
    }
    
    public void fly() {
        System.out.println("Flying Car is flying");
    }
}

class HydroFlyingCar extends FlyingCar {
    HydroFlyingCar(String name, String color, int engineHP) {
        super(name,color,engineHP);
    }
    
    public void driveInWater() {
        System.out.println("Hydro-flying car driving on the surface of the water");
    }
}

public class Main {
    public static void main(String[] args) {
        HydroFlyingCar hfc = new HydroFlyingCar("SuperFlySwim","Black",5000);
        hfc.drive();
        hfc.reverse();
        hfc.fly();
        hfc.driveInWater();
    }
}
